commit/merkleroot/rmn/client.go

```diff
@@ -19,13 +33,533 @@ type Client interface {
+) ([]rmnSignedObservationWithMeta, error) {
+	respChan := c.rawRmnClient.Recv()
+
+	finishedRequestIDs := mapset.NewSet[uint64]()
+	rmnObservationResponses := make([]rmnSignedObservationWithMeta, 0)
+	chainObservationsCount := make(map[uint64]int)
+
+	initialObservationRequestTimer := time.NewTimer(c.observationsInitialRequestTimerDuration)
+	defer initialObservationRequestTimer.Stop()
+	for {
+		select {
+		case resp := <-respChan:
+			parsedResp, err := c.parseResponse(&resp, requestIDs, finishedRequestIDs)
+			if err != nil {
+				c.lggr.Warnw("failed to parse RMN response", "err", err)
+				continue
+			}
+			finishedRequestIDs.Add(parsedResp.RequestId)
+
+			signedObs := parsedResp.GetSignedObservation()
+			if signedObs == nil {
+				c.lggr.Errorf("RMN node returned an unexpected type of response: %+v", parsedResp.Response)
+			} else {
+				err := c.validateSignedObservationResponse(resp.RMNNodeID, lursPerChain, signedObs, destChain)
+				if err != nil {
+					c.lggr.Errorw("failed to validate signed observation response", "err", err)
+					continue
+				}
+				rmnObservationResponses = append(rmnObservationResponses, rmnSignedObservationWithMeta{
+					RMNNodeID:         resp.RMNNodeID,
+					SignedObservation: signedObs,
+				})
+				for _, lu := range signedObs.Observation.FixedDestLaneUpdates {
+					chainObservationsCount[lu.LaneSource.SourceChainSelector]++
+				}
+			}
+
+			// We got all the responses we were waiting for.
+			if finishedRequestIDs.Equal(requestIDs) {
+				return rmnObservationResponses, nil
+			}
+
+			// We got sufficient number of observation responses for every source chain.
+			allChainsHaveEnoughResponses := true
+			for _, count := range chainObservationsCount {
+				if count < c.minObservers {
+					allChainsHaveEnoughResponses = false
+					break
+				}
+			}
+			if allChainsHaveEnoughResponses {
+				return rmnObservationResponses, nil
+			}
+		case <-initialObservationRequestTimer.C:
+			// Timer expired, send the observation requests to the rest of the RMN nodes.
+			requestsPerNode := make(map[uint32][]*rmnpb.FixedDestLaneUpdateRequest)
+			for sourceChain, lur := range lursPerChain {
+				for nodeID := range lur.rmnNodes.Iter() {
+					if requestedNodes[sourceChain].Contains(nodeID) {
+						continue
+					}
+					requestedNodes[sourceChain].Add(nodeID)
+					if _, ok := requestsPerNode[nodeID]; !ok {
+						requestsPerNode[nodeID] = make([]*rmnpb.FixedDestLaneUpdateRequest, 0)
+					}
+					requestsPerNode[nodeID] = append(requestsPerNode[nodeID], lur.lur)
+				}
+			}
+			newRequestIDs := c.sendObservationRequests(destChain, requestsPerNode)
+			requestIDs = requestIDs.Union(newRequestIDs)
+		case <-ctx.Done():
+			return nil, ErrTimeout
+		}
+	}
+}
+
+func (c *PBClient) validateSignedObservationResponse(
+	rmnNodeID uint32,
+	lurs map[uint64]lurWithMeta,
+	signedObs *rmnpb.SignedObservation,
+	destChain *rmnpb.LaneDest,
+) error {
+	if signedObs.Observation.LaneDest != destChain {
```

Comment:
Is this comparison correct? Seems to be comparing pointers which may not work as expected?